### PR TITLE
Upgrade to support Jupyter Lab 3

### PR DESCRIPTION
**Pull request by jacobtomlinson**
Upgrade to support Jupyter Lab 3 via the [migration guide](https://jupyterlab.readthedocs.io/en/stable/extension/extension_migration.html).

The primary benefit here other than just staying up to date is that the JavaScript extensions support modular prebuilt code. So whole extensions can now be distributed via pip/conda without needing nodejs or npm.

### CODE REVIEW

File: .github/workflows/build.yml
Comment (raydouglass) [inline .github/workflows/build.yml:1]: What's this file for? Does gpuCI not cover something?

**Review by raydouglass — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml
Comment (jacobtomlinson) [reply .github/workflows/build.yml:1]: I think this may be leftover from before gpuCI was added. Could probably be removed now.

Comment (jacobtomlinson) [reply .github/workflows/build.yml:1]: Actually I think this was created by the Jupyter Lab migration tool. So can definitely be deleted. But we may need to update gpuCI to match the functionality.

**Comment by rcthomas** [comment]
It looks like applying the migration tool to this labextension works pretty smoothly, but that CI issues are blocking the merge, is that correct?

**Comment by jacobtomlinson** [comment]
@rcthomas yeah I think so.

@raydouglass could you help me get this unblocked. I'm not sure what the CI is unhappy about.

**Comment by ajschmidt8** [comment]
@jacobtomlinson, I was able to replicate the build errors from [this Jenkins log](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/jupyterlab-nvdashboard/job/prb/job/jupyterlab-nvdashboard-cpu-build/39/console) locally. Ultimately, it seems the `jupyter labextension build .` command shown below is exiting with an error code of `1`.

image


After running `jupyter labextension build . -h`, the output (below) seems to indicate that `build` may not be a valid subcommand. Any thoughts on this? Do you think another package may need to be installed in order to enable the `build` subcommand?

<details>

<summary>`jupyter labextension build . -h` output</summary>

```
Work with JupyterLab extensions

Subcommands
===========
Subcommands are launched as `jupyter labextension cmd [args]`. For information
on using subcommand 'cmd', do: `jupyter labextension cmd -h`.

install
    Install labextension(s)
update
    Update labextension(s)
uninstall
    Uninstall labextension(s)
list
    List labextensions
link
    Link labextension(s)
unlink
    Unlink labextension(s)
enable
    Enable labextension(s)
disable
    Disable labextension(s)
check
    Check labextension(s)

Options
=======
The options below are convenience aliases to configurable class-options,
as listed in the "Equivalent to" description-line of the aliases.
To see all configurable class-options for some <cmd>, use:
    <cmd> --help-all

--debug
    set log level to logging.DEBUG (maximize logging output)
    Equivalent to: [--Application.log_level=10]
--generate-config
    generate default config file
    Equivalent to: [--JupyterApp.generate_config=True]
-y
    Answer yes to any questions instead of prompting.
    Equivalent to: [--JupyterApp.answer_yes=True]
--log-level=<Enum>
    Set the log level by value or name.
    Choices: any of [0, 10, 20, 30, 40, 50, 'DEBUG', 'INFO', 'WARN', 'ERROR', 'CRITICAL']
    Default: 30
    Equivalent to: [--Application.log_level]
--config=<Unicode>
    Full path of a config file.
    Default: ''
    Equivalent to: [--JupyterApp.config_file]

Examples
--------

    jupyter labextension list                        # list all configured labextensions
    jupyter labextension install <extension name>    # install a labextension
    jupyter labextension uninstall <extension name>  # uninstall a labextension

To see all available configurables, use `--help-all`.

```

</details>

**Comment by jacobtomlinson** [comment]
Thanks for digging into this @ajschmidt8.

This PR was created mostly by an automated migration tool from Jupyter folks. Obviously that tool doesn't support migrating GPUCI scripts, so I'm guessing something is wrong in there and will need migrating by hand.

Digging into it I think we want `jlpm build` now.

I'll make some updates to this PR and see how it goes.

**Comment by jacobtomlinson** [comment]
@rcthomas do you have any thoughts on why we are seeing this error about `jupyter labextension build`?

**Comment by rcthomas** [comment]
@jacobtomlinson No I don't, let me page @marius311 here who I know has built and tested the labextension himself.  If the call to labextension build was added by the migration tool I think he would have hit it since he informed me that the tool was all he needed to use.

**Comment by marius311** [comment]
> let me page @marius311 here who I know has built and tested the labextension himself

Correct, my exact sequence of commands was
```
git clone <this repo>
python -m jupyterlab.upgrade_extension .
pip install -e .
jlpm install
jlpm run build
jupyter labextension install .
```
and it works. 

This is the output of `jupyter labextension build` if it helps in any way, but I probably don't know enough to help beyond that. 


```
$ jupyter labextension build
Building extension in /global/u1/m/marius/src/jupyterlab-nvdashboard

Compilation starting…


Compilation finished

asset 568.eefb815965ad9d435529.js?v=eefb815965ad9d435529 9.52 KiB [emitted] [immutable] [minimized]
asset remoteEntry.849945296e2fd9727bd4.js 6.55 KiB [emitted] [immutable] [minimized] (name: jupyterlab-nvdashboard)
asset 534.706d853a65bc86cdc1b8.js?v=706d853a65bc86cdc1b8 3.34 KiB [emitted] [immutable] [minimized]
runtime modules 17.6 KiB 12 modules
orphan modules 325 bytes [orphan] 1 module
built modules 20.7 KiB (javascript) 336 bytes (consume-shared) 42 bytes (share-init) [built]
  javascript modules 20.7 KiB
    modules by path ./style/ 4.5 KiB 6 modules
    modules by path ./node_modules/ 9.04 KiB 3 modules
    modules by path ./lib/*.js 7.06 KiB 2 modules
    container entry 42 bytes [built] [code generated]
  consume-shared-module modules 336 bytes
    modules by path consume shared module (default) @jupyterlab/ 168 bytes 4 modules
    modules by path consume shared module (default) @lumino/ 84 bytes 2 modules
    modules by path *.1 (singleton) 84 bytes
      consume shared module (default) react@^17.0.1 (singleton) 42 bytes [built] [code generated]
      consume shared module (default) react-dom@^17.0.1 (singleton) 42 bytes [built] [code generated]
  provide shared module (default) jupyterlab-nvdashboard@0.5.0 = ./lib/index.js 42 bytes [built] [code generated]
webpack 5.28.0 compiled successfully in 3205 ms
```

**Comment by jacobtomlinson** [comment]
Thansk @marius311. Any idea why we are seeing errors in CI that `build` is not a valid subcommand of `jupyter labextension`? Are we missing some dependency?

**Comment by txu0393** [comment]
Hey folks, just spitballing here but maybe the reason why the CI can't find the `build` subcommand is because it needs to have JupyterLab 3 installed?
From the logs, it seems like it still has JupyterLab 2.X

**Comment by jacobtomlinson** [comment]
It looks like the automatic migration tool added some GitHub Actions stuff. I've tried to roll that into the GPUCI scripts.

I've also tried to migrate the packaging over following the docs on [this page](https://jupyterlab.readthedocs.io/en/stable/extension/extension_tutorial.html#extension-tutorial-publish).

I've successfully managed to build the wheel locally and then install it into a fresh conda environment and everything works as expected. So no more NPM packging 🎉.

I'm not 100% sure if anything needs changing in the conda recipe or if things are correct in the CI scripts.

@ajschmidt8 would you mind reviewing?

Also @jakirkham if you have some time I would really appreciate a packaging review here. Jupyter Lab 3 changed things a bunch. Specifically that we need to use `build` to compile the JavaScript and bundle it, so I'm not sure how that maps over to conda recipes.

File: conda/recipes/jupyterlab-nvdashboard/meta.yaml
Comment (jakirkham) [inline conda/recipes/jupyterlab-nvdashboard/meta.yaml:19]:
```suggestion
  host:
```

File: conda/recipes/jupyterlab-nvdashboard/meta.yaml
Comment (jakirkham) [inline conda/recipes/jupyterlab-nvdashboard/meta.yaml:20]: Should we drop Python 3.6? Asking as RAPIDS and Dask have now dropped it

File: conda/recipes/jupyterlab-nvdashboard/meta.yaml
Comment (jakirkham) [inline conda/recipes/jupyterlab-nvdashboard/meta.yaml:28]: As a precaution :)

```suggestion
    - jupyterlab >=3.0.0,<4
```

**Review by jakirkham — COMMENTED** [review]
I'm not too familiar with the changes needed for JL 3 and Conda, but made a few suggestions based on scraping recipes in conda-forge

File: conda/recipes/jupyterlab-nvdashboard/meta.yaml
Comment (jakirkham) [inline conda/recipes/jupyterlab-nvdashboard/meta.yaml:35]: Also we may want [some file tests like these]( https://github.com/conda-forge/jupytext-feedstock/pull/54/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR39-R42 )

**Comment by ajschmidt8** [comment]
@jacobtomlinson, I updated a few things in CI here and fixed some linting errors. The `jupyter serverextension list 2>&1 | grep -ie "jupyterlab_nvdashboard.*OK"` command [here](https://github.com/jacobtomlinson/jupyterlab-nvdashboard/blob/f0c92b75d189e76ed91e0da0ed2cfc06ced6a6b1/ci/gpu/build.sh#L66) is currently exiting with a non-zero exit code. I've commented it out in a2b5251. Can it be removed? The `jupyter labextension list` command [below it](https://github.com/jacobtomlinson/jupyterlab-nvdashboard/blob/f0c92b75d189e76ed91e0da0ed2cfc06ced6a6b1/ci/gpu/build.sh#L67) passes. I'm not too familiar with this plugin/repo, but I'm wondering if it's a `labextension` and _not_ a `serverextension` and that's why the command fails.


image

File: conda/recipes/jupyterlab-nvdashboard/meta.yaml
Comment (jacobtomlinson) [reply conda/recipes/jupyterlab-nvdashboard/meta.yaml:20]: Yeah that probably makes sense

**Comment by jacobtomlinson** [comment]
Thanks for the review @jakirkham.

Thanks @ajschmidt8, this extension has both a serverextension and a labextension so both checks should pass.

**Comment by jacobtomlinson** [comment]
Actually @ajschmidt8 I think you're right. The server-side component in this extension uses the `jupyter_serverproxy_servers` entrypoint to start the server component, not a server extension.

File: conda/recipes/jupyterlab-nvdashboard/meta.yaml
Comment (jacobtomlinson) [reply conda/recipes/jupyterlab-nvdashboard/meta.yaml:35]: Thanks, added in 255fa71

**Comment by jacobtomlinson** [comment]
Ok I think this is ready for a merge. @ajschmidt8 could you take one last look?

**Review by ajschmidt8 — COMMENTED** [review]
I've opened up https://github.com/rapidsai/integration/pull/261 to update the version of JupyterLab used in our containers. Once that PR is merged and the packages & containers are rebuilt and published, we can remove the lines that I've marked here since our containers:

- already have `nodejs` installed ([link](https://github.com/rapidsai/integration/blob/b58391925bdffbb7c00639dbae521ef8054f0414/conda/recipes/versions.yaml#L103-L104))
- will have the updated `jupyterlab` version installed (via PR above)
- `jupyter-packaging` is a dependency of `jupyterlab` ([link](https://anaconda.org/conda-forge/jupyterlab/files))

Let's wait till the above PR is merged & published before removing these lines since CI will fail if we remove them before then.

**Comment by txu0393** [comment]
Are there any updates with this PR?

**Comment by ajschmidt8** [comment]
rerun tests

**Comment by ajschmidt8** [comment]
> Are there any updates with this PR?

@txu0393, assuming CI passes during this next run, we should be all set to merge.

**Comment by Atharex** [comment]
@ajschmidt8 you might want to review the build script. It seems it's using a faulty command 
https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fjupyterlab-nvdashboard%2Fprb%2Fjupyterlab-nvdashboard-cpu-build/detail/jupyterlab-nvdashboard-cpu-build/53/pipeline#log-933

```
  $ jupyter labextension build .
  Please supply at least one subcommand: check, disable, enable, install, link, list, uninstall, unlink, update
  error Command failed with exit code 1.
```

**Comment by Atharex** [comment]
This looks like a failure due to different jupyterlab versions...

The centos pipeline is explicitly installing jupyterlab v3 
https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fjupyterlab-nvdashboard%2Fprb%2Fjupyterlab-nvdashboard-gpu-build/detail/jupyterlab-nvdashboard-gpu-build/166/pipeline/#log-230 

The failing ubuntu pipeline seems to be using only jupyterlab 2 
https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fjupyterlab-nvdashboard%2Fprb%2Fjupyterlab-nvdashboard-gpu-build/detail/jupyterlab-nvdashboard-gpu-build/166/pipeline/#log-230

Other ubuntu pipelines again explicitly install jupyterlab 3 
https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fjupyterlab-nvdashboard%2Fprb%2Fjupyterlab-nvdashboard-gpu-build/detail/jupyterlab-nvdashboard-gpu-build/165/pipeline#log-245

https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fjupyterlab-nvdashboard%2Fprb%2Fjupyterlab-nvdashboard-gpu-build/detail/jupyterlab-nvdashboard-gpu-build/164/pipeline#log-217

https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fjupyterlab-nvdashboard%2Fprb%2Fjupyterlab-nvdashboard-gpu-build/detail/jupyterlab-nvdashboard-gpu-build/166/pipeline/#log-230

**Comment by ajschmidt8** [comment]
There was an upstream PR for CI to resolve this that was opened but never merged. I just merged it. Let's try this again.

rerun tests

**Comment by jacobtomlinson** [comment]
Looks like CI passed 🎉

**Comment by txu0393** [comment]
Woot! Hoping to see this released soon!!!

**Review by ajschmidt8 — APPROVED** [review]
LGTM.

**Comment by jacobtomlinson** [comment]
@txu0393 we are aiming to release this week.

**Pull request merged**
